### ISSUE TITLE
Add an option to remove only consecutive history duplicates

### ISSUE THREAD

**Issue by segevfiner** [PowerShell/PSReadLine#750]
Environment data
----------------

```
PS version: 5.1.17134.228
PSReadline version: 1.2
os: 10.0.17134.1 (WinBuild.160101.0800)
PS file version: 10.0.17134.1 (WinBuild.160101.0800)
```

Description
------------
PSReadline has the option `Set-PSReadlineOption -HistoryNoDuplicates` (The plain Windows console has the equivalent "Discard Old Duplicates"), this option removes *all* duplicate history entries, even non-consecutive ones. For example: Typing the commands `echo 1`, `echo 2`, `echo 2`, `echo 1`, the history will be `echo 1`, `echo 2`.

Bash has the option `HISTCONTROL=ignoredups`, and Zsh has `HIST_IGNORE_DUPS`. These options only remove *consecutive* duplicates. For example: Typing the commands `echo 1`, `echo 2`, `echo 2`, `echo 1`, the history will be `echo 1`, `echo 2`, `echo 1`.

I think the Bash/Zsh behavior is way more intuitive and useful than removing all duplicates and it would be a great feature for PSReadline to have.

**EDIT:** There is also the related `HISTCONTROL=ignorespace`, `HIST_IGNORE_SPACE` which is also nice to have and probably quite easy to implement.

**Comment by lzybkr** [comment]
If the Bash/Zsh option was named `ignore_consecutive_dupes`, then I might agree it's more intuitive :)

There has been some iteration on how `-HistoryNoDuplicates` behaves, for reference, see https://github.com/lzybkr/PSReadLine/issues/562 and https://github.com/lzybkr/PSReadLine/issues/74,

You might try [2.0.0-beta2](https://www.powershellgallery.com/packages/PSReadLine/2.0.0-beta2) and see if that is more intuitive (it includes #562).

I have considered a (optional) richer UI while going through history, see https://github.com/lzybkr/PSReadLine/issues/107 for some thoughts.

**Comment by lzybkr** [comment]
HIST_IGNORE_SPACE is easy - you can add something like this to your profile:

```PowerShell
Set-PSReadLineOption -AddToHistoryHandler {
    param([string]$line)
    return $line.Length -gt 3 -and $line[0] -ne ' ' -and $line[0] -ne ';'
}
```

This example will ignore short commands (3 or less) and commands that start with a space or semi-colon.

I thought I had this example in the sample profile but apparently not.

**Comment by daxian-dbw** [comment]
> For example: Typing the commands `echo 1`, `echo 2`, `echo 2`, `echo 1`, the history will be `echo 1`, `echo 2`.

The reported issue has been addressed in PSReadLine (I guess it was done quite a while back). Now `HistoryNoDuplicates` only removes consecutive duplicates.

![image](https://user-images.githubusercontent.com/127450/139141094-d15e7350-3366-459c-b5b6-29cb45ee8b5d.png)

**Comment by segevfiner** [comment]
Any idea in which version? Os that version bundled with PowerShell yet?

**Comment by daxian-dbw** [comment]
The screenshot above is what I got when using the 2.1.0 version of PSReadLine. The fix may be in the 2.0.0 version as well, but I didn't check.

The 2.1.0 version is bundled with PowerShell 7.1.x. You can also get it from PowerShell Gallery.

**Comment by lzybkr** [comment]
I could be mistaken, but I think the very first version released in Windows 10 probably worked the same.
